File: data_baseline/gpt-5/java/rep8/Main.java
import com.sun.net.httpserver.*;
import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.time.*;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import com.google.gson.*;
import com.google.gson.reflect.TypeToken;

public class Main {
    static class User {
        int id;
        String username;
        String passwordHash; // sha-256
        String salt;
        User(int id, String username, String passwordHash, String salt) {
            this.id = id;
            this.username = username;
            this.passwordHash = passwordHash;
            this.salt = salt;
        }
    }

    static class Todo {
        int id;
        int userId;
        String title;
        String description;
        boolean completed;
        String created_at;
        String updated_at;
    }

    static class Database {
        static final AtomicInteger nextUserId = new AtomicInteger(1);
        static final AtomicInteger nextTodoId = new AtomicInteger(1);
        static final Map<Integer, User> usersById = new ConcurrentHashMap<>();
        static final Map<String, User> usersByUsername = new ConcurrentHashMap<>();
        static final Map<String, Integer> sessions = new ConcurrentHashMap<>();
        static final Map<Integer, Todo> todosById = new ConcurrentHashMap<>();
    }

    static final Gson gson = new GsonBuilder().serializeNulls().create();
    static final DateTimeFormatter ISO_INSTANT_SECONDS = DateTimeFormatter.ISO_INSTANT; // Instant already in Z

    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i = 0; i < args.length; i++) {
            if ("--port".equals(args[i]) && i + 1 < args.length) {
                port = Integer.parseInt(args[i + 1]);
                i++;
            }
        }
        HttpServer server = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
        server.createContext("/", new RootHandler());
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("Server started on 0.0.0.0:" + port);
    }

    static class RootHandler implements HttpHandler {
        @Override
        public void handle(HttpExchange exchange) throws IOException {
            try {
                String method = exchange.getRequestMethod();
                URI uri = exchange.getRequestURI();
                String path = uri.getPath();

                // Routing
                if (method.equals("POST") && path.equals("/register")) {
                    handleRegister(exchange);
                    return;
                }
                if (method.equals("POST") && path.equals("/login")) {
                    handleLogin(exchange);
                    return;
                }
                if (method.equals("POST") && path.equals("/logout")) {
                    handleLogout(exchange);
                    return;
                }
                if (method.equals("GET") && path.equals("/me")) {
                    handleMe(exchange);
                    return;
                }
                if (method.equals("PUT") && path.equals("/password")) {
                    handlePassword(exchange);
                    return;
                }
                if (path.equals("/todos") && method.equals("GET")) {
                    handleTodosList(exchange);
                    return;
                }
                if (path.equals("/todos") && method.equals("POST")) {
                    handleTodosCreate(exchange);
                    return;
                }
                if (path.startsWith("/todos/") && path.length() > "/todos/".length()) {
                    String idStr = path.substring("/todos/".length());
                    Integer id = null;
                    try {
                        id = Integer.parseInt(idStr);
                    } catch (NumberFormatException nfe) {
                        sendJsonError(exchange, 404, "Todo not found");
                        return;
                    }
                    switch (method) {
                        case "GET":
                            handleTodoGet(exchange, id);
                            return;
                        case "PUT":
                            handleTodoUpdate(exchange, id);
                            return;
                        case "DELETE":
                            handleTodoDelete(exchange, id);
                            return;
                        default:
                            break;
                    }
                }

                // Not found
                sendJsonError(exchange, 404, "Not found");
            } catch (Exception e) {
                // Internal server error
                e.printStackTrace();
                sendJsonError(exchange, 500, "Internal server error");
            } finally {
                try { exchange.close(); } catch (Exception ignore) {}
            }
        }
    }

    // Helper utilities
    static String readBody(HttpExchange exchange) throws IOException {
        InputStream is = exchange.getRequestBody();
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] buf = new byte[4096];
        int r;
        while ((r = is.read(buf)) != -1) baos.write(buf, 0, r);
        return baos.toString(StandardCharsets.UTF_8);
    }

    static void setJsonContentType(HttpExchange exchange) {
        Headers h = exchange.getResponseHeaders();
        h.set("Content-Type", "application/json");
    }

    static void sendJson(HttpExchange exchange, int statusCode, Object obj) throws IOException {
        setJsonContentType(exchange);
        byte[] bytes = gson.toJson(obj).getBytes(StandardCharsets.UTF_8);
        exchange.sendResponseHeaders(statusCode, bytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(bytes);
        }
    }

    static void sendJsonError(HttpExchange exchange, int statusCode, String message) throws IOException {
        Map<String, String> err = new LinkedHashMap<>();
        err.put("error", message);
        sendJson(exchange, statusCode, err);
    }

    static Integer authenticate(HttpExchange exchange) throws IOException {
        String cookieHeader = exchange.getRequestHeaders().getFirst("Cookie");
        if (cookieHeader == null) return null;
        String sessionId = null;
        String[] parts = cookieHeader.split(";\\s*");
        for (String part : parts) {
            int eq = part.indexOf('=');
            if (eq > 0) {
                String name = part.substring(0, eq).trim();
                String val = part.substring(eq + 1).trim();
                if (name.equals("session_id")) {
                    sessionId = val;
                    break;
                }
            }
        }
        if (sessionId == null) return null;
        Integer userId = Database.sessions.get(sessionId);
        return userId;
    }

    static void requireAuthOr401(HttpExchange exchange) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null || Database.usersById.get(userId) == null) {
            sendJsonError(exchange, 401, "Authentication required");
            throw new RuntimeException("unauthorized");
        }
    }

    static String isoNow() {
        Instant now = Instant.now().truncatedTo(ChronoUnit.SECONDS);
        return ISO_INSTANT_SECONDS.format(now);
    }

    static String sha256(String s) {
        try {
            java.security.MessageDigest md = java.security.MessageDigest.getInstance("SHA-256");
            byte[] b = md.digest(s.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte x : b) sb.append(String.format("%02x", x));
            return sb.toString();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    static String generateSalt() {
        byte[] b = new byte[16];
        new java.security.SecureRandom().nextBytes(b);
        StringBuilder sb = new StringBuilder();
        for (byte x : b) sb.append(String.format("%02x", x));
        return sb.toString();
    }

    static boolean verifyPassword(User u, String password) {
        String hash = sha256(u.salt + ":" + password);
        return hash.equals(u.passwordHash);
    }

    static String cleanString(JsonObject obj, String key) {
        if (obj == null || !obj.has(key) || obj.get(key).isJsonNull()) return null;
        return obj.get(key).getAsString();
    }

    static Boolean getBoolean(JsonObject obj, String key) {
        if (obj == null || !obj.has(key) || obj.get(key).isJsonNull()) return null;
        try { return obj.get(key).getAsBoolean(); } catch (Exception e) { return null; }
    }

    // Handlers
    static void handleRegister(HttpExchange exchange) throws IOException {
        String body = readBody(exchange);
        JsonObject req;
        try {
            req = JsonParser.parseString(body).getAsJsonObject();
        } catch (Exception e) {
            sendJsonError(exchange, 400, "Invalid JSON");
            return;
        }
        String username = cleanString(req, "username");
        String password = cleanString(req, "password");
        if (username == null || !username.matches("^[a-zA-Z0-9_]{3,50}$")) {
            sendJsonError(exchange, 400, "Invalid username");
            return;
        }
        if (password == null || password.length() < 8) {
            sendJsonError(exchange, 400, "Password too short");
            return;
        }
        // uniqueness
        synchronized (Database.usersByUsername) {
            if (Database.usersByUsername.containsKey(username)) {
                sendJsonError(exchange, 409, "Username already exists");
                return;
            }
            int id = Database.nextUserId.getAndIncrement();
            String salt = generateSalt();
            String hash = sha256(salt + ":" + password);
            User u = new User(id, username, hash, salt);
            Database.usersById.put(id, u);
            Database.usersByUsername.put(username, u);
            Map<String, Object> resp = new LinkedHashMap<>();
            resp.put("id", id);
            resp.put("username", username);
            sendJson(exchange, 201, resp);
        }
    }

    static void handleLogin(HttpExchange exchange) throws IOException {
        String body = readBody(exchange);
        JsonObject req;
        try {
            req = JsonParser.parseString(body).getAsJsonObject();
        } catch (Exception e) {
            sendJsonError(exchange, 401, "Invalid credentials");
            return;
        }
        String username = cleanString(req, "username");
        String password = cleanString(req, "password");
        if (username == null || password == null) {
            sendJsonError(exchange, 401, "Invalid credentials");
            return;
        }
        User u = Database.usersByUsername.get(username);
        if (u == null || !verifyPassword(u, password)) {
            sendJsonError(exchange, 401, "Invalid credentials");
            return;
        }
        String token = UUID.randomUUID().toString();
        Database.sessions.put(token, u.id);
        Headers h = exchange.getResponseHeaders();
        h.add("Set-Cookie", "session_id=" + token + "; Path=/; HttpOnly");
        Map<String, Object> resp = new LinkedHashMap<>();
        resp.put("id", u.id);
        resp.put("username", u.username);
        sendJson(exchange, 200, resp);
    }

    static void handleLogout(HttpExchange exchange) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) {
            sendJsonError(exchange, 401, "Authentication required");
            return;
        }
        String cookieHeader = exchange.getRequestHeaders().getFirst("Cookie");
        if (cookieHeader != null) {
            String[] parts = cookieHeader.split(";\\s*");
            for (String part : parts) {
                int eq = part.indexOf('=');
                if (eq > 0) {
                    String name = part.substring(0, eq).trim();
                    String val = part.substring(eq + 1).trim();
                    if (name.equals("session_id")) {
                        Database.sessions.remove(val);
                    }
                }
            }
        }
        Map<String, Object> resp = new LinkedHashMap<>();
        sendJson(exchange, 200, resp);
    }

    static void handleMe(HttpExchange exchange) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) {
            sendJsonError(exchange, 401, "Authentication required");
            return;
        }
        User u = Database.usersById.get(userId);
        if (u == null) {
            sendJsonError(exchange, 401, "Authentication required");
            return;
        }
        Map<String, Object> resp = new LinkedHashMap<>();
        resp.put("id", u.id);
        resp.put("username", u.username);
        sendJson(exchange, 200, resp);
    }

    static void handlePassword(HttpExchange exchange) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) {
            sendJsonError(exchange, 401, "Authentication required");
            return;
        }
        User u = Database.usersById.get(userId);
        if (u == null) {
            sendJsonError(exchange, 401, "Authentication required");
            return;
        }
        String body = readBody(exchange);
        JsonObject req;
        try {
            req = JsonParser.parseString(body).getAsJsonObject();
        } catch (Exception e) {
            sendJsonError(exchange, 400, "Invalid JSON");
            return;
        }
        String oldp = cleanString(req, "old_password");
        String newp = cleanString(req, "new_password");
        if (oldp == null || !verifyPassword(u, oldp)) {
            sendJsonError(exchange, 401, "Invalid credentials");
            return;
        }
        if (newp == null || newp.length() < 8) {
            sendJsonError(exchange, 400, "Password too short");
            return;
        }
        String newSalt = generateSalt();
        String newHash = sha256(newSalt + ":" + newp);
        u.salt = newSalt;
        u.passwordHash = newHash;
        Map<String, Object> resp = new LinkedHashMap<>();
        sendJson(exchange, 200, resp);
    }

    static void handleTodosList(HttpExchange exchange) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) { sendJsonError(exchange, 401, "Authentication required"); return; }
        List<Todo> out = new ArrayList<>();
        for (Todo t : Database.todosById.values()) {
            if (t.userId == userId) out.add(t);
        }
        out.sort(Comparator.comparingInt(a -> a.id));
        sendJson(exchange, 200, out);
    }

    static void handleTodosCreate(HttpExchange exchange) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) { sendJsonError(exchange, 401, "Authentication required"); return; }
        String body = readBody(exchange);
        JsonObject req;
        try { req = JsonParser.parseString(body).getAsJsonObject(); }
        catch (Exception e) { sendJsonError(exchange, 400, "Invalid JSON"); return; }
        String title = cleanString(req, "title");
        String description = cleanString(req, "description");
        if (title == null || title.trim().isEmpty()) {
            sendJsonError(exchange, 400, "Title is required");
            return;
        }
        if (description == null) description = "";
        Todo t = new Todo();
        t.id = Database.nextTodoId.getAndIncrement();
        t.userId = userId;
        t.title = title;
        t.description = description;
        t.completed = false;
        t.created_at = isoNow();
        t.updated_at = t.created_at;
        Database.todosById.put(t.id, t);
        sendJson(exchange, 201, t);
    }

    static Todo findOwnedTodo(int id, int userId) {
        Todo t = Database.todosById.get(id);
        if (t == null) return null;
        if (t.userId != userId) return null;
        return t;
    }

    static void handleTodoGet(HttpExchange exchange, int id) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) { sendJsonError(exchange, 401, "Authentication required"); return; }
        Todo t = findOwnedTodo(id, userId);
        if (t == null) { sendJsonError(exchange, 404, "Todo not found"); return; }
        sendJson(exchange, 200, t);
    }

    static void handleTodoUpdate(HttpExchange exchange, int id) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) { sendJsonError(exchange, 401, "Authentication required"); return; }
        Todo t = findOwnedTodo(id, userId);
        if (t == null) { sendJsonError(exchange, 404, "Todo not found"); return; }
        String body = readBody(exchange);
        JsonObject req;
        try { req = JsonParser.parseString(body).getAsJsonObject(); }
        catch (Exception e) { sendJsonError(exchange, 400, "Invalid JSON"); return; }
        if (req.has("title")) {
            String title = cleanString(req, "title");
            if (title == null || title.trim().isEmpty()) {
                sendJsonError(exchange, 400, "Title is required");
                return;
            }
            t.title = title;
        }
        if (req.has("description")) {
            String desc = cleanString(req, "description");
            t.description = (desc == null ? "" : desc);
        }
        if (req.has("completed")) {
            Boolean comp = getBoolean(req, "completed");
            if (comp != null) t.completed = comp.booleanValue();
        }
        t.updated_at = isoNow();
        sendJson(exchange, 200, t);
    }

    static void handleTodoDelete(HttpExchange exchange, int id) throws IOException {
        Integer userId = authenticate(exchange);
        if (userId == null) { sendJsonError(exchange, 401, "Authentication required"); return; }
        Todo t = findOwnedTodo(id, userId);
        if (t == null) { sendJsonError(exchange, 404, "Todo not found"); return; }
        Database.todosById.remove(id);
        // 204 No Content, no body
        exchange.sendResponseHeaders(204, -1);
    }
}

File: data_baseline/gpt-5/scala/rep8/Main.scala
//> using scala "3.3.1"

import java.net.InetSocketAddress
import com.sun.net.httpserver.{HttpExchange, HttpHandler, HttpServer}
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.UUID
import java.util.concurrent.{ConcurrentHashMap, Executors}
import java.util.concurrent.atomic.AtomicInteger
import scala.jdk.CollectionConverters._

object Main {
  // Models
  case class UserRec(id: Int, username: String, password: String)
  case class TodoRec(
      id: Int,
      userId: Int,
      title: String,
      description: String,
      completed: Boolean,
      createdAt: String,
      updatedAt: String
  )

  // In-memory storage
  object Store {
    val usersById = new ConcurrentHashMap[Int, UserRec]()
    val userIdByUsername = new ConcurrentHashMap[String, Integer]()
    val todosById = new ConcurrentHashMap[Int, TodoRec]()
    val sessions = new ConcurrentHashMap[String, Integer]() // token -> userId
    val nextUserId = new AtomicInteger(1)
    val nextTodoId = new AtomicInteger(1)
  }

  // Helpers
  def nowIso(): String = Instant.now().truncatedTo(java.time.temporal.ChronoUnit.SECONDS).toString

  def readBody(is: InputStream): String = {
    val buf = new Array[Byte](8192)
    val sb = new StringBuilder
    var n = is.read(buf)
    while (n != -1) {
      sb.append(new String(buf, 0, n, StandardCharsets.UTF_8))
      n = is.read(buf)
    }
    sb.toString
  }

  def jsonEscape(s: String): String =
    s.flatMap {
      case '"' => "\\\""
      case '\\' => "\\\\"
      case '\n' => "\\n"
      case '\r' => "\\r"
      case '\t' => "\\t"
      case c if c.isControl => f"\\u${c.toInt}%04x"
      case c => c.toString
    }

  def sendJson(exchange: HttpExchange, code: Int, body: String): Unit = {
    val bytes = body.getBytes(StandardCharsets.UTF_8)
    val headers = exchange.getResponseHeaders
    headers.set("Content-Type", "application/json")
    exchange.sendResponseHeaders(code, bytes.length)
    val os = exchange.getResponseBody
    try os.write(bytes)
    finally os.close()
  }

  def sendNoContent(exchange: HttpExchange): Unit = {
    exchange.sendResponseHeaders(204, -1) // no body
    exchange.getResponseBody.close()
  }

  def jsonError(exchange: HttpExchange, code: Int, msg: String): Unit = {
    val body = s"{" + s"\"error\":\"${jsonEscape(msg)}\"" + "}"
    sendJson(exchange, code, body)
  }

  def parseCookie(header: String): Map[String, String] = {
    if (header == null) Map.empty
    else header.split(';').toList.flatMap { part =>
      val kv = part.trim.split("=", 2)
      if (kv.length == 2) Some(kv(0) -> kv(1)) else None
    }.toMap
  }

  def getSessionUser(exchange: HttpExchange): Option[UserRec] = {
    val cookieHeader = exchange.getRequestHeaders.getFirst("Cookie")
    val cookies = parseCookie(cookieHeader)
    cookies.get("session_id").flatMap { token =>
      val uidOpt = Option(Store.sessions.get(token)).map(_.intValue())
      uidOpt.flatMap { uid => Option(Store.usersById.get(uid)) }
    }
  }

  def setSessionCookie(exchange: HttpExchange, token: String): Unit = {
    exchange.getResponseHeaders.add("Set-Cookie", s"session_id=$token; Path=/; HttpOnly")
  }

  def parseJsonFieldStr(body: String, key: String): Option[String] = {
    val pattern = ("\\\"" + java.util.regex.Pattern.quote(key) + "\\\"\\s*:\\s*\\\"(.*?)\\\"").r
    pattern.findFirstMatchIn(body).map(_.group(1))
  }
  def parseJsonFieldBool(body: String, key: String): Option[Boolean] = {
    val pattern = ("\\\"" + java.util.regex.Pattern.quote(key) + "\\\"\\s*:\\s*(true|false)").r
    pattern.findFirstMatchIn(body).map(m => m.group(1) == "true")
  }

  def validateUsername(u: String): Boolean = u.matches("^[a-zA-Z0-9_]+$") && u.length >= 3 && u.length <= 50
  def validatePassword(p: String): Boolean = p != null && p.length >= 8

  def handle(exchange: HttpExchange): Unit = {
    try {
      val method = exchange.getRequestMethod
      val path = exchange.getRequestURI.getPath

      (method, path) match {
        case ("POST", "/register") =>
          val body = readBody(exchange.getRequestBody)
          val usernameOpt = parseJsonFieldStr(body, "username")
          val passwordOpt = parseJsonFieldStr(body, "password")
          (usernameOpt, passwordOpt) match {
            case (Some(username), Some(password)) =>
              if (!validateUsername(username)) jsonError(exchange, 400, "Invalid username")
              else if (!validatePassword(password)) jsonError(exchange, 400, "Password too short")
              else {
                val newId = Store.nextUserId.get()
                val existed = Option(Store.userIdByUsername.putIfAbsent(username, Integer.valueOf(newId))).isDefined
                if (existed) jsonError(exchange, 409, "Username already exists")
                else {
                  Store.nextUserId.incrementAndGet()
                  Store.usersById.put(newId, UserRec(newId, username, password))
                  val resp = s"{" + s"\"id\":$newId,\"username\":\"${jsonEscape(username)}\"" + "}"
                  sendJson(exchange, 201, resp)
                }
              }
            case _ => jsonError(exchange, 400, "Invalid JSON")
          }

        case ("POST", "/login") =>
          val body = readBody(exchange.getRequestBody)
          (parseJsonFieldStr(body, "username"), parseJsonFieldStr(body, "password")) match {
            case (Some(username), Some(password)) =>
              val uidOpt = Option(Store.userIdByUsername.get(username)).map(_.intValue())
              val userOpt = uidOpt.flatMap(id => Option(Store.usersById.get(id)))
              userOpt match {
                case Some(u) if u.password == password =>
                  val token = UUID.randomUUID().toString.replaceAll("-", "")
                  Store.sessions.put(token, Integer.valueOf(u.id))
                  setSessionCookie(exchange, token)
                  val resp = s"{" + s"\"id\":${u.id},\"username\":\"${jsonEscape(u.username)}\"" + "}"
                  sendJson(exchange, 200, resp)
                case _ => jsonError(exchange, 401, "Invalid credentials")
              }
            case _ => jsonError(exchange, 400, "Invalid JSON")
          }

        case ("POST", "/logout") =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(_) =>
              val cookieHeader = exchange.getRequestHeaders.getFirst("Cookie")
              val cookies = parseCookie(cookieHeader)
              cookies.get("session_id").foreach(tok => Store.sessions.remove(tok))
              sendJson(exchange, 200, "{}")
          }

        case ("GET", "/me") =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val resp = s"{" + s"\"id\":${u.id},\"username\":\"${jsonEscape(u.username)}\"" + "}"
              sendJson(exchange, 200, resp)
          }

        case ("PUT", "/password") =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val body = readBody(exchange.getRequestBody)
              (parseJsonFieldStr(body, "old_password"), parseJsonFieldStr(body, "new_password")) match {
                case (Some(oldp), Some(newp)) =>
                  if (u.password != oldp) jsonError(exchange, 401, "Invalid credentials")
                  else if (!validatePassword(newp)) jsonError(exchange, 400, "Password too short")
                  else {
                    Store.usersById.put(u.id, u.copy(password = newp))
                    sendJson(exchange, 200, "{}")
                  }
                case _ => jsonError(exchange, 400, "Invalid JSON")
              }
          }

        case (meth, p) if p == "/todos" && meth == "GET" =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val todos = Store.todosById.values().asScala.toList.filter(_.userId == u.id).sortBy(_.id)
              val arr = todos.map { t =>
                s"{" +
                  s"\"id\":${t.id}," +
                  s"\"title\":\"${jsonEscape(t.title)}\"," +
                  s"\"description\":\"${jsonEscape(t.description)}\"," +
                  s"\"completed\":${t.completed}," +
                  s"\"created_at\":\"${jsonEscape(t.createdAt)}\"," +
                  s"\"updated_at\":\"${jsonEscape(t.updatedAt)}\"" +
                "}"
              }.mkString(",")
              sendJson(exchange, 200, s"[${arr}]")
          }

        case (meth, p) if p == "/todos" && meth == "POST" =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val body = readBody(exchange.getRequestBody)
              val titleOpt = parseJsonFieldStr(body, "title").map(_.trim)
              val desc = parseJsonFieldStr(body, "description").getOrElse("")
              titleOpt match {
                case Some(t) if t.nonEmpty =>
                  val id = Store.nextTodoId.getAndIncrement()
                  val now = nowIso()
                  val rec = TodoRec(id, u.id, t, desc, false, now, now)
                  Store.todosById.put(id, rec)
                  val resp = s"{" +
                    s"\"id\":${rec.id}," +
                    s"\"title\":\"${jsonEscape(rec.title)}\"," +
                    s"\"description\":\"${jsonEscape(rec.description)}\"," +
                    s"\"completed\":${rec.completed}," +
                    s"\"created_at\":\"${jsonEscape(rec.createdAt)}\"," +
                    s"\"updated_at\":\"${jsonEscape(rec.updatedAt)}\"" +
                  "}"
                  sendJson(exchange, 201, resp)
                case _ => jsonError(exchange, 400, "Title is required")
              }
          }

        case (meth, p) if p.startsWith("/todos/") && meth == "GET" =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val idStr = p.stripPrefix("/todos/")
              val idOpt = idStr.toIntOption
              idOpt match {
                case None => jsonError(exchange, 404, "Todo not found")
                case Some(id) =>
                  val t = Store.todosById.get(id)
                  if (t == null || t.userId != u.id) jsonError(exchange, 404, "Todo not found")
                  else {
                    val resp = s"{" +
                      s"\"id\":${t.id}," +
                      s"\"title\":\"${jsonEscape(t.title)}\"," +
                      s"\"description\":\"${jsonEscape(t.description)}\"," +
                      s"\"completed\":${t.completed}," +
                      s"\"created_at\":\"${jsonEscape(t.createdAt)}\"," +
                      s"\"updated_at\":\"${jsonEscape(t.updatedAt)}\"" +
                    "}"
                    sendJson(exchange, 200, resp)
                  }
              }
          }

        case (meth, p) if p.startsWith("/todos/") && meth == "PUT" =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val idStr = p.stripPrefix("/todos/")
              idStr.toIntOption match {
                case None => jsonError(exchange, 404, "Todo not found")
                case Some(id) =>
                  val rec = Store.todosById.get(id)
                  if (rec == null || rec.userId != u.id) jsonError(exchange, 404, "Todo not found")
                  else {
                    val body = readBody(exchange.getRequestBody)
                    val titleOpt = parseJsonFieldStr(body, "title").map(_.trim)
                    titleOpt match {
                      case Some(t) if t.isEmpty => jsonError(exchange, 400, "Title is required")
                      case _ =>
                        val descOpt = parseJsonFieldStr(body, "description")
                        val compOpt = parseJsonFieldBool(body, "completed")
                        val updated = rec.copy(
                          title = titleOpt.getOrElse(rec.title),
                          description = descOpt.getOrElse(rec.description),
                          completed = compOpt.getOrElse(rec.completed),
                          updatedAt = nowIso()
                        )
                        Store.todosById.put(id, updated)
                        val resp = s"{" +
                          s"\"id\":${updated.id}," +
                          s"\"title\":\"${jsonEscape(updated.title)}\"," +
                          s"\"description\":\"${jsonEscape(updated.description)}\"," +
                          s"\"completed\":${updated.completed}," +
                          s"\"created_at\":\"${jsonEscape(updated.createdAt)}\"," +
                          s"\"updated_at\":\"${jsonEscape(updated.updatedAt)}\"" +
                        "}"
                        sendJson(exchange, 200, resp)
                    }
                  }
              }
          }

        case (meth, p) if p.startsWith("/todos/") && meth == "DELETE" =>
          getSessionUser(exchange) match {
            case None => jsonError(exchange, 401, "Authentication required")
            case Some(u) =>
              val idStr = p.stripPrefix("/todos/")
              idStr.toIntOption match {
                case None => jsonError(exchange, 404, "Todo not found")
                case Some(id) =>
                  val rec = Store.todosById.get(id)
                  if (rec == null || rec.userId != u.id) jsonError(exchange, 404, "Todo not found")
                  else { Store.todosById.remove(id); sendNoContent(exchange) }
              }
          }

        case _ =>
          jsonError(exchange, 404, "Not found")
      }
    } catch {
      case _: Throwable =>
        try jsonError(exchange, 500, "Internal server error")
        catch case _: Throwable => ()
    } finally {
      if (exchange.getRequestBody != null) try exchange.getRequestBody.close() catch { case _: Throwable => () }
    }
  }

  def parsePort(args: Array[String], default: Int = 8080): Int = {
    val arr = args.toList
    arr.sliding(2, 1).collectFirst { case List("--port", p) if p.forall(_.isDigit) => p.toInt }.getOrElse(default)
  }

  def main(args: Array[String]): Unit = {
    val port = parsePort(args)
    val server = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0)
    server.createContext("/", new HttpHandler { def handle(exchange: HttpExchange): Unit = Main.handle(exchange) })
    val exec = Executors.newFixedThreadPool(Math.max(4, Runtime.getRuntime.availableProcessors()))
    server.setExecutor(exec)
    server.start()
    // Block forever
    println(s"Server started on 0.0.0.0:$port")
    val latch = new Object()
    latch.synchronized { latch.wait() }
  }
}
